workflow-api/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    pub async fn timer(&self, _: Duration) {
+        todo!()
+    }
+}
+
+pub struct SignalDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The handler input type must be erased here, since otherwise we couldn't store/return the
+    // heterogeneous collection of definition types in the workflow itself. The signal macro
+    // will wrap the user's function with code that performs deserialization.
+    handler: Box<dyn FnMut(&mut WF, Payload) -> Result<(), anyhow::Error>>,
+}
+pub struct QueryDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The query macro will wrap the user's function with code that performs deserialization of
+    // input and serialization of output, as well as error boxing.
+    handler: Box<dyn FnMut(&WF, Payload) -> Result<Payload, anyhow::Error>>,
+}
+
+/// TODO: Placeholders, likely belong inside protos crate. These will be auto-implemented for
+///   anything using serde already (which I expect is how virtually everyone will do this).
+pub trait TemporalSerializable {}
+impl<T> TemporalSerializable for T {}
+pub trait TemporalDeserializable {}
+impl<T> TemporalDeserializable for T {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::FutureExt;
+    use std::{collections::HashMap, marker::PhantomData};
+
+    // Workflow implementation example
+    struct MyWorkflow {
+        foo: u64,
+        bar: HashMap<String, u64>,
+    }
+
+    impl Workflow for MyWorkflow {
+        type Input = String;
+        type Output = u64;
+        const NAME: &'static str = "MyWorkflowType";
+
+        fn new(input: Self::Input, _ctx: SafeWfContext) -> Self {
+            let mut bar = HashMap::new();
+            bar.insert(input, 10);
+            Self { foo: 0, bar }
+        }
+
+        fn run(
+            &mut self,
+            ctx: WfContext,
+        ) -> BoxFuture<Result<WfExitValue<Self::Output>, anyhow::Error>> {
+            async move {
+                ctx.timer(Duration::from_secs(1)).await;
+                self.foo = 1;
+                // See activity definitions file
+                ctx.my_act_fn("Hi!").await.unwrap();
+                // The into() is unfortunately unavoidable without making C-A-N and confirm cancel
+                // be errors instead. Personally, I don't love that and I think it's not idiomatic
+                // Rust, whereas needing to `into()` something is. Other way would be macros, but
+                // it's slightly too much magic I think.
+                Ok(self.foo.into())
+            }
+            .boxed()
+            // TODO: The need to box here is slightly unfortunate, but it's either that or require
+            //  users to depend on `async_trait` (which just hides the same thing). IMO this is the
+            //  best option until more language features stabilize and this can go away.
```

Comment:
Can you just require `run` be an `async fn`?

Reply:
Nah, that's the thing that can't work in a trait yet.

Reply:
May make it worth doing it attribute style. PyO3 are heavy users of this approach over trait-based (e.g. https://pyo3.rs/v0.18.3/class and https://pyo3.rs/v0.18.3/function).

Reply:
That's just async trait at that point, but if it ends up being useful for other reasons I'm definitely OK with a procmacro

Reply:
Maybe something like:

```rust
#[workflow]
struct MyWorkflow {}

impl MyWorkflow {
    #[workflow_run]
    async fn run(&mut self, some_arg: String) -> TemporalResult<Strnig> {
        // ...
    }
    
    async fn my_signal(&mut self, some_arg) {
        // ...
    }
}
```

Reply:
(and a `#[workflow_init]` or `#[workflow_new]` or whatever named attribute on a func that accepts workflow args to create the struct if it's not "default"...we built this attribute for .NET constructors to allow for readonly field initialization)

Reply:
The `#[workflow]` attribute would just go over the impl block I think. There's no need for it on the struct definition, and that allows you to just require a `new` and a `run` method without them needing separate attributes.

Reply:
Nice. May be worth exploring at proposal time if there aren't too many drawbacks. If caught between two impls that both may kinda work, just list pros/cons of each and then start one of our classic design Zoom calls to pick a winner :-)

Reply:
Mega necropost

Preparing for pres at offsite... I'm now pretty convinced `#[workflow]` attribute macro is the best way to go for sure. Fixes `async` problem, makes signals/queries automatically work, enforces `new` & `run`... It's the way to go. People who want to could still do it all by hand if they hate it for whatever reason

Reply:
And it's so pretty!